R/data.table.R

```diff
@@ -1892,73 +1890,23 @@ as.matrix.data.table = function(x, rownames=NULL, rownames.value=NULL, ...) {
                " which is outside the column number range [1,ncol=", ncol(x), "].")
       }
     }
-  } else if (!is.null(rownames.value)) {
-    if (length(rownames.value)!=nrow(x))
-      stop("length(rownames.value)==", length(rownames.value),
-           " but should be nrow(x)==", nrow(x))
   }
-  if (!is.null(rownames)) {
-    # extract that column and drop it.
-    rownames.value = x[[rownames]]
-    dm = dim(x) - 0:1
-    cn = names(x)[-rownames]
-    X = x[, .SD, .SDcols = cn]
-  } else {
-    dm = dim(x)
-    cn = names(x)
-    X = x
-  }
-  if (any(dm == 0L))
-    return(array(NA, dim = dm, dimnames = list(rownames.value, cn)))
-  p = dm[2L]
-  n = dm[1L]
-  collabs = as.list(cn)
+  
+  # Create shallow copy - where each element of the list X is simply a pointer to
+  # the same column in the data.table x. This means that we do not modify the 
+  # input data.table when dropping rows or coercing columns.
```

Comment:
please include test number for that here in the comment

Reply:
On my TODO list. Will plan to add a suite of tests to cover the various possible cases in the C code as well that make the assumption that updating a column pointer makes a copy and does not modify the input data.table.